server/src/main/java/org/apache/druid/client/indexing/ClientCompactionIntervalSpec.java

```diff
@@ -44,12 +45,27 @@
   @Nullable
   private final String sha256OfSortedSegmentIds;
 
-  public static ClientCompactionIntervalSpec fromSegments(List<DataSegment> segments)
+  public static ClientCompactionIntervalSpec fromSegments(List<DataSegment> segments, Granularity segmentGranularity)
   {
+    Interval interval = JodaUtils.umbrellaInterval(segments.stream().map(DataSegment::getInterval).collect(Collectors.toList()));
+    if (segmentGranularity != null) {
+      // If segmentGranularity is set, then the segmentGranularity of the segments may not align with the configured segmentGranularity
+      // We must adjust the interval of the compaction task to fully cover and align with the segmentGranularity
+      // For example,
+      // - The umbrella interval of the segments is 2015-04-11/2015-04-12 but configured segmentGranularity is YEAR,
+      // if the compaction task's interval is 2015-04-11/2015-04-12 then we can run into race condition where after submitting
+      // the compaction task, a new segment outside of the interval (i.e. 2015-02-11/2015-02-12) got created will be lost as it is
+      // overshadowed by the compacted segment (compacted segment has interval 2015-01-01/2016-01-01.
+      // Hence, in this case, we must adjust the compaction task interval to 2015-01-01/2016-01-01.
+      // - The umbrella interval of the segments is 2015-02-01/2015-03-01 but configured segmentGranularity is MONTH,
+      // if the compaction task's interval is 2015-02-01/2015-03-01 then compacted segments created will be
+      // 2015-01-26/2015-02-02, 2015-02-02/2015-02-09, 2015-02-09/2015-02-16, 2015-02-16/2015-02-23, 2015-02-23/2015-03-02.
+      // The compacted segment would cause existing data from 2015-01-26 to 2015-02-01 and 2015-03-01 to 2015-03-02 to be lost.
+      //  Hence, in this case, we must adjust the compaction task interval to 2015-01-26/2015-03-02
+      interval = JodaUtils.umbrellaInterval(segmentGranularity.getIterable(interval));
```

Comment:
Looking at `JodaUtils.umbrellaInterval()`, I think this code can cause OOM if `segmentGranularity.getIterable()` returns lots of intervals. `JodaUtils.umbrellaInterval()` doesn't seem to need to store all `startDates` and `endDates` in memory. Instead, it can keep only one pair of `minStartDate` and `maxEndDate` which are updated in the loop. Should we fix it in this PR too?

Reply:
Or should we have a guardrail for `segmentGranularity` to not return such many intervals?

Reply:
Done

Reply:
I don't think we need a guardrail here. The list of segments passed to this method should be from a single time chunk (from the CompactionSegmentIterator). The single time chuck is then adjusted based on some segmentGranularity. Since the original interval (from the umbrellaInterval of the list of segments) is limited to a single time chunk, segmentGranularity.getIterable() should not returns lots of intervals. The edge case to this is if the granularities are either ALL or NONE. I think we should put a guardrail against ALL or NONE but not here. (actually I think it would blow up in the NewestSegmentFirstIterator before getting here anyway)

Reply:
I agree that the guardrail should be somewhere else than here.